Utility type `KnownKeys<T>` broken on TypeScript 4.3

### Description

The following code no longer compiles after upgrading from TypeScript 4.2.4 to 4.3.2

<details>
<summary>(Click here to expand)</summary>

```ts
import { App } from "@slack/bolt";
const app = new App();
app.command("/say_hi", async ({ ack, respond }) => {
    await ack();
    await respond({
        blocks: [{ type: "section", text: { text: "hi", type: "plain_text" } }],
    //  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    // error TS2345: Argument of type '{ blocks: { type: string; text: { text: string; type: string; }; }[]; response_type: "ephemeral"; }' is not assignable to parameter of type 'string | RespondArguments'.
    //  Object literal may only specify known properties, and 'blocks' does not exist in type 'RespondArguments'.    });
        response_type: "ephemeral",
    });
});
```

</details>

I have narrowed down the problem to the utility type `KnownKeys<T>` in `/src/types/helpers.ts`.

```ts
type type KnownKeys<T> = {
    [K in keyof T]: string extends K ? never : number extends K ? never : K;
} extends { [_ in keyof T]: infer U }
    ? U
    : never;

type Foo = {
    foo: "Foo";
    bar: "Bar";
    [key: string]: any;
};

type Result = KnownKeys<Foo>;
// type of `Result` is `never` in TypeScript 4.3.2
// type of `Result` is `"foo" | "bar"` in TypeScript 4.2.4
```

`KnownKeys<T>` failing to resolve the keys has caused `RespondArguments` in `/src/types/utilities.ts` to omit the property "blocks" by calling `Pick` with `never` as the 2nd type argument:

```ts
type RespondArguments = Pick<
  ChatPostMessageArguments,
  Exclude<KnownKeys<ChatPostMessageArguments>, 'channel' | 'text'> // <-- KnownKeys<T> fails here
> & { /* ... */ };
```

...which in turn causes `RespondFn` to report an error when "blocks" is included in `respond({ /* ... */ })`

### What type of issue is this? (place an `x` in one of the `[ ]`)
- [x] bug
- [ ] enhancement (feature request)
- [ ] question
- [ ] documentation related
- [ ] example code related
- [ ] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [x] I've read and understood the [Contributing guidelines](https://github.com/slackapi/bolt/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

---

### Bug Report

Filling out the following details about bugs will help us solve your issue sooner.

#### Reproducible in:

package version: 3.3.0

node version: 15.12.0

OS version(s): OSX 11.3.1

#### Steps to reproduce:

(See description)

Also, `/types-tests/utilities.test-d.ts` should show an error after upgrading to TypeScript 4.3 (See attachment)

#### Expected result:

`KnownKeys<T>` should work properly in TypeScript 4.3.  
Maybe add a test case to cover `type Result = KnownKeys<Foo>;` should not infer type `never`.

#### Actual result:

`KnownKeys<T>` infers `never` in TypeScript 4.3.

#### Attachments:

![image](https://user-images.githubusercontent.com/61226932/120292084-8f7cdb80-c2f6-11eb-9c6f-2c62d20f09a3.png)

--- a/src/types/helpers.ts
+++ b/src/types/helpers.ts
@@ -4,13 +4,10 @@
 export type StringIndexed = Record<string, any>;
 
 /**
- * Type function which removes the index signature for the type `T`
+ * @deprecated No longer works in TypeScript 4.3
  */
-export type KnownKeys<T> = {
-  [K in keyof T]: string extends K ? never : number extends K ? never : K;
-} extends { [_ in keyof T]: infer U }
-  ? U
-  : never;
+// @ts-ignore -- ignore unused type parameter warning
+export type KnownKeys<T> = never;
 
 /**
  * Type function which allows either types `T` or `U`, but not both.

--- a/src/types/utilities.ts
+++ b/src/types/utilities.ts
@@ -1,9 +1,27 @@
 import { ChatPostMessageArguments, WebAPICallResult } from '@slack/web-api';
-import { KnownKeys } from './helpers';
+
+// (issue#951) KnownKeys<ChatPostMessageArguments> no longer works in TypeScript 4.3
+type ChatPostMessageArgumentsKnownKeys =
+  | 'token'
+  | 'channel'
+  | 'text'
+  | 'as_user'
+  | 'attachments'
+  | 'blocks'
+  | 'icon_emoji'
+  | 'icon_url'
+  | 'link_names'
+  | 'mrkdwn'
+  | 'parse'
+  | 'reply_broadcast'
+  | 'thread_ts'
+  | 'unfurl_links'
+  | 'unfurl_media'
+  | 'username';
 
 // The say() utility function binds the message to the same channel as the incoming message that triggered the
 // listener. Therefore, specifying the `channel` argument is not required.
-export type SayArguments = Pick<ChatPostMessageArguments, Exclude<KnownKeys<ChatPostMessageArguments>, 'channel'>> & {
+export type SayArguments = Pick<ChatPostMessageArguments, Exclude<ChatPostMessageArgumentsKnownKeys, 'channel'>> & {
   channel?: string;
 };
 
@@ -13,7 +31,7 @@ export interface SayFn {
 
 export type RespondArguments = Pick<
   ChatPostMessageArguments,
-  Exclude<KnownKeys<ChatPostMessageArguments>, 'channel' | 'text'>
+  Exclude<ChatPostMessageArgumentsKnownKeys, 'channel' | 'text'>
 > & {
   /** Response URLs can be used to send ephemeral messages or in-channel messages using this argument */
   response_type?: 'in_channel' | 'ephemeral';

